components/patient-data/rest/src/main/java/org/phenotips/data/rest/internal/DefaultPatientsFetchResourceImpl.java

```diff
@@ -0,0 +1,222 @@
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        }
+    }
+
+    /**
+     * Retrieves patient entities given a list of external patient IDs.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param eids a list of external patient IDs, as strings
+     */
+    private void addEids(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final List<Object> eids)
+    {
+        for (final Object eid : eids) {
+            if (StringUtils.isNotBlank((String) eid)) {
+                collectPatientsFromEid(patientsBuilder, eid);
+            }
+        }
+    }
+
+    /**
+     * Retrieves and collects patient entities that correspond to the provided external ID.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param eid an external patient ID, as string
+     */
+    private void collectPatientsFromEid(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final Object eid)
+    {
+        try {
+            final Query q = qm.createQuery("where doc.object(PhenoTips.PatientClass).external_id = :eid", Query.XWQL);
+            q.bindValue("eid", eid);
+            final List<Object> patientIds = q.execute();
+            addIds(patientsBuilder, patientIds);
+        } catch (final QueryException ex) {
+            logger.warn("Failed to retrieve patient with external id [{}]: {}", eid, ex.getMessage());
+        }
+    }
+
+    /**
+     * Retrieves patient entities given a list of internal patient IDs.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param ids a list of patient ids, as strings
+     */
+    private void addIds(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final List<Object> ids)
+    {
+        for (final Object id : ids) {
+            if (StringUtils.isNotBlank((String) id)) {
+                addPatientFromId(patientsBuilder, id);
+            }
+        }
+    }
+
+    /**
+     * Given the patient's internal ID, retrieves the patient entity, if it exists and if the user has view rights, and
+     * adds it to the set of patient entities.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param id an internal patient ID
+     */
+    private void addPatientFromId(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final Object id)
+    {
+        try {
+            // Try to get the patient entity.
+            final PrimaryEntity patient = this.repository.get((String) id);
+            // Get the current user.
+            final User currentUser = this.users.getCurrentUser();
+            // If user has view rights and patient with the provided ID exists, add patient to patient set.
+            if (patient != null && this.access.hasAccess(Right.VIEW, currentUser == null ? null
```

Comment:
Instead of manually checking rights here, you should `@Inject @Named("secure") PatientRepository`, which will throw a `SecurityException` when trying to get a patient that's not accessible.